### PR TITLE
Feat/directives validation

### PR DESCRIPTION
**Pull request by lheagy**
## Validate Directives

Add validation on directives and directives List as per #578. 
- `BaseDirective` has a method `validate(self, directiveList=None)` which returns `True` by default
-  `DirectiveList` has a method `validate(self)` that goes through the directive list and validates each one
- `InvProb` runs the directive list validation on setter

### CODE REVIEW

File: SimPEG/Directives.py [hunk 1]
```diff
@@ -588,6 +597,28 @@ def endIter(self):
...
+    def validate(self, directiveList):
+        # check if a linear preconditioner is in the list, if not warn else
+        # assert that it is listed after the IRLS directive
+        dList = directiveList.dList
+        self_ind = dList.index(self)
+        lin_precond_ind = [
+            isinstance(d, Update_lin_PreCond) for d in dList
+        ]
+
+        if any(lin_precond_ind):
```
Comment (lheagy) [inline SimPEG/Directives.py]: @fourndo: can you take a look at this? Here I check if the there is a `Update_lin_PreCond` directive in the directive list and then make sure it is after the IRLS directive or throw a warning if not in the list (not an error - see what you think is more appropriate)

File: SimPEG/Inversion.py [hunk 2]
```diff
@@ -21,6 +21,7 @@ class BaseInversion(object):
     def directiveList(self):
         if getattr(self, '_directiveList', None) is None:
             self._directiveList = Directives.DirectiveList(inversion=self)
+            self._directiveList.validate()  # validate if we skip setter
```
Comment (lheagy) [inline SimPEG/Inversion.py]: @rowanc1: what do you think of these locations for the validation?

File: examples/05-mag/plot_MVI_Cartesian.py [hunk 3]
```diff
@@ -174,7 +174,7 @@ def run(plotIt=True):
     update_Jacobi = Directives.Update_lin_PreCond()
 
     inv = Inversion.BaseInversion(invProb,
-                                  directiveList=[betaest, update_Jacobi, IRLS, ])
+                                  directiveList=[betaest, IRLS, update_Jacobi])
```
Comment (lheagy) [inline examples/05-mag/plot_MVI_Cartesian.py]: @fourndo: I updated your example :)

File: SimPEG/Directives.py (hunk 1 above)
Comment (lheagy) [inline SimPEG/Directives.py]: you can check out the test file that I added to see it in action

File: requirements_dev.txt [hunk 4]
```diff
@@ -14,3 +14,4 @@ ipython
 matplotlib
 properties[math]
 wheel
+pytest
```
Comment (lheagy) [inline requirements_dev.txt]: @rowanc1: this is to test that we are raising warnings

**Comment by lheagy** [comment]
@fourndo, @rowanc1: this is a fairly light one. Please have a look when you have a chance and let me know what you think! @fourndo: this is coming in on the mag amplitude branch, so I will let you merge once both you and @rowanc1 are happy :)

**Comment by lheagy** [comment]
@fourndo: I went through and resolved the merge conflicts here. If you are happy with these changes, can you please merge this pr. Thanks!

**Comment by lheagy** [comment]
This will also update MagAmplitude to the most recent ref/objectivefunctions

**Comment by fourndo** [comment]
Yup sounds great @lheagy , I ll keep an eye out for the end of the test.

**Comment by lheagy** [comment]
@fourndo: it looks like there are quite a number of other changes on the MagAmplitde branch that are making getting this pr to pass a bit tricky. So I am going to close this and re-open a pr onto the objective function branch with the directives changes. You can then merge that upstream to MagAmplitude.

**Pull request closed without merging**